src/Windows/dll/local_cache.cpp

```diff
@@ -1,33 +1,412 @@
-    time_t expiry,
-    size_t data_size,
-    const void* data)
+static void throw_errno(const std::string& description)
 {
-    (void)id;
-    (void)expiry;
-    (void)data_size;
-    (void)data;
+    throw_errno(description, errno);
+}
+
+struct CacheEntryHeaderV1 {
+    uint16_t version;   // The version of the cache header
+    time_t expiry;      // expiration time of this cache item
+};
+
+static void make_dir(const std::string& dirname)
+{
+    struct _stat buf {};
+    int rc = _stat(dirname.c_str(), &buf);
+
+    if (rc == 0)
+    {
+        if ((buf.st_mode & _S_IFDIR) > 0)
+        {
+            return;
+        }
+
+        throw std::runtime_error(dirname + " already exists, and is not a directory.");
+    }
+
+    wchar_t dirName[64];
+    std::mbstowcs(dirName, dirname.c_str(), strlen(dirname.c_str()) + 1);//Plus null
+
+    rc = CreateDirectory(dirName, NULL);
+    if (rc != 0)
+    {
+        throw_errno("Error creating directory '" + dirname + "'");
+    }
+}
+
+static void init_callback()
+{
+    const char * env_home = ::getenv("APPDATA");
+    const char * env_azdcap_cache = ::getenv("AZDCAP_CACHE");
+    const std::string application_name("\\.az-dcap-client");
+
+    std::string dirname;
+
+    if (env_azdcap_cache != 0 && (strcmp(env_azdcap_cache, "") != 0))
+    {
+        dirname = env_azdcap_cache;
+    }
+    else if (env_home != 0 && (strcmp(env_home, "") != 0))
+    {
+        dirname = std::string(env_home);
+    }
+    else
+    {
+        // Throwing exception if the expected HOME
+        // environment variable is not defined.
+        throw std::runtime_error("HOME and AZDCAPCACHE environment variables not defined");
+    }
+
+    dirname += application_name;
+    make_dir(dirname);
+    g_cache_dirname = dirname;
+}
+
+
+static void init()
+{
+    static std::once_flag init_flag;
+    std::call_once(init_flag, init_callback);
+}
+
+static std::string sha256(size_t data_size, const void* data)
+{
+    BCRYPT_ALG_HANDLE hAlg = NULL;
+    BCRYPT_HASH_HANDLE hHash = NULL;
+    NTSTATUS status = STATUS_UNSUCCESSFUL;
+    std::string errorString;
+    DWORD cbData = 0;
+    DWORD cbHash = 0;
+    DWORD cbHashObject = 0;
+    PBYTE pbHashObject = NULL;
+    PBYTE pbHash = NULL;
+    std::string retval;
+
+    //open an algorithm handle
+    if (!NT_SUCCESS(status = BCryptOpenAlgorithmProvider(
+        &hAlg,
+        BCRYPT_SHA256_ALGORITHM,
+        NULL,
+        0)))
+    {
+        errorString = "Error 0x" + std::to_string(status) + "returned by BCryptOpenAlgorithmProvider\n";
+        goto Cleanup;
+    }
+
+    if (!NT_SUCCESS(status == BCryptGetProperty(
+        hAlg,
+        BCRYPT_OBJECT_LENGTH,
+        (PBYTE)&cbHashObject,
+        sizeof(DWORD),
+        &cbData,
+        0)))
+    {
+        errorString = "Error 0x" + std::to_string(status) + "returned by BCryptGetProperty\n";
+        goto Cleanup;
+    }
+
+    pbHashObject = (PBYTE)HeapAlloc(GetProcessHeap(), 0, cbHashObject);
```

Comment:
Why not use std::unique_ptr<BYTE> here? That means you don't have to manually free the buffer.

Reply:
Used vector